extensions/amp-bind/amp-bind.md

```diff
@@ -219,32 +304,165 @@ Only binding to the following components and attributes are allowed. Most bindab
+
+## Appendix
+
+### `<amp-state>` specification
+
+An `amp-state` element may contain either a child `<script>` element **OR** a `src` attribute containing a CORS URL to a remote JSON endpoint, but not both.
+
+```html
+<amp-state id="myLocalState">
+  <script type="application/json">
+    {
+      "foo": "bar"
+    }
+  </script>
+</amp-state>
+
+<amp-state id="myRemoteState" src="https://data.com/articles.json">
+</amp-state>
+```
+
+#### Attributes
+
+**src**
+
+The URL of the remote endpoint that will return the JSON that will update this `amp-state`. This must be a CORS HTTP service.
+
+The `src` attribute allows all standard URL variable substitutions. See the [Substitutions Guide](../../spec/amp-var-substitutions.md) for more info.
+
+**credentials** (optional)
+
+Defines a `credentials` option as specified by the [Fetch API](https://fetch.spec.whatwg.org/).
+To send credentials, pass the value of "include". If this is set, the response must follow
+the [AMP CORS security guidelines](../../spec/amp-cors-requests.md).
+
+The support values are "omit" and "include". Default is "omit".
+
+### Custom Built-in Functions
+
+`amp-bind` includes the following builtin functions.
+
+| Name | Description | Arguments | Examples |
+| --- | --- | --- | --- |
+| `copyAndSplice` | Similar to [Array#splice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) except a copy of the spliced array is returned. | `array` : An array.<br>`start` : Index at which to start changing the array.<br> `deleteCount` : The number of items to delete, starting at index `start`<br> `items...` Items to add to the array, beginning at index `start`. | `// Deleting an element. Returns [1, 3]`<br>`copyAndSplice([1, 2, 3], 1, 1)`<br>`// Replacing an item. Returns ['Pizza', 'Cake', 'Ice Cream']`<br>`copyAndSplice(['Pizza', 'Cake', 'Soda'], 2, 1, 'Ice Cream')` |
+
+### Deep-merge with `AMP.setState()`
+
+When `AMP.setState()` is called `amp-bind` deep-merges the provided object literal with the current state. All variables from the object literal are written to the state directly except for nested objects, which are recursively merged.
+
+Consider the following example:
+
+```javascript
+{
+<!-- State is empty -->
+}
+```
+
+```html
+<button on="tap:AMP.setState({employee: {name: 'John Smith', age: 47, vehicle: 'Car'}})"...></button>
+<button on="tap:AMP.setState({employee: {age: 64}})"...></button>
+```
+
+When the first button is pressed, the state changes to:
+
+```javascript
+{
+  employee: {
+    name: 'John Smith',
+    age: 47,
+    vehicle: 'Car',
+  }
+}
+```
+
+When the second button is pressed, `amp-bind` will **deep-merge** this state with the state introduced by the button press.
+
+```javascript
+{
+  employee: {
+    name: 'John Smith',
+    age: 64,
+    vehicle: 'Car',
+  }
+}
+```
+
```

Comment:
Mention somewhere that arrays and primitives are always overwritten, and an error will be thrown if the provided object has circular references.